test/e2e/dataplane/service_discovery.go

```diff
@@ -122,19 +121,23 @@ func verifyClusterIpWithDig(f *framework.Framework, cluster framework.ClusterInd
 }
 
 func verifyCurl(f *framework.Framework, cluster framework.ClusterIndex, pod v1.Pod, target string, expectingSuccess bool) {
+
 	cmd := []string{"curl", target}
-	stdout, _, err := f.ExecWithOptions(framework.ExecOptions{
-		Command:       cmd,
-		Namespace:     f.Namespace,
-		PodName:       pod.Name,
-		ContainerName: pod.Spec.Containers[0].Name,
-		CaptureStdout: true,
-		CaptureStderr: true,
-	}, framework.ClusterB)
-	if expectingSuccess {
-		Expect(err).NotTo(HaveOccurred())
-		Expect(stdout).To(ContainSubstring("Welcome to nginx!"))
-	} else {
-		Expect(err).To(HaveOccurred())
-	}
+	By(fmt.Sprintf("Executing %q to verify service %q is discoverable, expectingSuccess to be %t", strings.Join(cmd, " "), target, expectingSuccess))
+	framework.AwaitUntil("verify if service is discoverable", func() (interface{}, error) {
+		stdout, _, err := f.ExecWithOptions(framework.ExecOptions{
+			Command:       cmd,
+			Namespace:     f.Namespace,
+			PodName:       pod.Name,
+			ContainerName: pod.Spec.Containers[0].Name,
+			CaptureStdout: true,
+			CaptureStderr: true,
+		}, cluster)
+		if (err != nil && expectingSuccess) || (err == nil && !expectingSuccess) {
```

Comment:
Won't this condition give a false positive?
`(err == nil && !expectingSuccess)`

 i.e. if err= nil then it'll return nil as the error, and youll get a success even though you were were expecting it to fail

Reply:
> Won't this condition give a false positive?
> `(err == nil && !expectingSuccess)`
> 
> i.e. if err= nil then it'll return nil as the error, and youll get a success even though you were were expecting it to fail

Yes, will fix this.